developing/cicd/README.md

```diff
@@ -8,9 +8,175 @@ actually be used by anyone. This section provides resources for setting
+
+### Test
+
+We will not be writing deeply about testing methodologies or specifics around kinds of tests and the philosophies here.
+
+In the context of a CI pipeline:
+
+**Tests should be hermetic.** Ideally, multiple CI runs should be able to run at the same time without affecting each other.
+
+**Tests should be idempotent.** Tests in CI should be able to run several times over without producing inconsistent effects.
+
+**Tests should be reproducible in a developer's computer.** So developers can debug them more easily.
+
+**Tests should run fast.** "Fast" is poorly defined here.
+Generally speaking, you want tests to be fast so merging code changes is fast.
+A developer shouldn't think they can start on a new piece of work in the hour or two to push a change t hrough CI and context switch.
+
+**Tests should fail fast.** Similar to the previous point, "fast" is relative but in CI you will want to give the waiting devloper signal on whether their work can be merged or requires changes as quick as possible.
+
+### Deployment
+
+Deployment will differ project to project and should take into account the requirements of the system being deployed to.
+
+**Deployment should be repeatable.** The same code and configuration should be deployable again and again.
+
+**Deployment process for software should be the same regardless of environment.** Using the same process and tooling means lower cognitive overhead in deployment and your non production environments should reflect production more closely.
+
+**Deployment tools should be maintained separate from the application being deployed.** We've seen this happen where we need to fix some deployed code but also had a problem with the code that performs the deployment and all of it is a hairy mess of debugging.
+
+### Release
+
+There are a few different ways of implementing release.
+No matter what, release is the step when you allow customers to have access to that particular version of your code.
+
+If you're releasing a CLI, it may mean that you're releasing your binary to a Homebrew tap or just a zip file to GitHub releases.
+
+If you have a service, you could have a loadbalancer manage traffic between different versions of your services.
+Or you could use something like Launch Darkly to leverage feature flags to do something similar for different user cohorts.
+
+## Continous (Integration | Delivery | Deploy)
+
+We throw around terms like CI or CD with the assumption that we know what the differences are between these practices.
+
+For clarity, at Truss we'll refer to them as follows:
+
+### Continuous Integration
+
+`Continuous Integration` is when developers merge changes into mainline as often as possible.
+Automated build and test steps validate changes and are incorporated into review before merging to mainline.
```

Comment:
Strictly speaking, CI is the practice of validating mainline on every push.  It doesn't really say anthing about branches. However, what you write is Best Practice as it means finding failure before we get to mainline. Ideally, mainline should never fail because that blocks changes from being merged. This is where a tool like [bors](https://bors.tech/) becomes interesting. It's probably worth talking about the value of keeping mainline failure free. Let me know if you want me to wax more on that.

Reply:
Made some edits to better reflect your views here. I think we might want more info on the value of keeping mainline failure free but I think that's maybe more detail than I was going for in this PR.